### PR TITLE
feat: add deterministic filler safe cuts

### LINKED ISSUE #42: [Feature] Implement deterministic filler detection and safe cut resolution
## User need

A filler-removal Skill needs deterministic semantic candidates and physically safe deletion ranges. The agent must not choose frame-level cut boundaries from natural-language reasoning.

## Proposed behavior

Add two editor-independent runtime components:

### FillerDetector

Convert revision-bound speech analysis into deterministic filler candidates using:

- Analyzer-provided filler metadata
- Configured filler vocabulary
- Confidence threshold
- Requested target range
- Protected speech evidence

Each candidate includes a preview-scoped stable ID, word evidence, confidence, threshold, source range, sequence range, occurrence identity, and structured reason codes.

### SafeCutResolver

Resolve candidate words into frame-aligned deletion ranges using:

- Word timestamps
- Speech/VAD segments
- Silence boundaries
- Neighboring words
- Protected breath or laughter segments
- Occurrence and target bounds
- Sequence frame duration
- Configured pause-preservation policy

The resolver returns `AUTO_APPLY`, `SUGGESTED`, or `SKIPPED` together with its evidence and optional `ripple-delete` operation.

This component performs no mutation.

### In scope

- Deterministic filler classification
- Stable preview-scoped candidate IDs
- Confidence and vocabulary policies
- Safe boundary resolution
- Frame alignment
- Protected segment handling
- Structured decision and skip reasons
- Pure unit and property-style tests

### Out of scope

- Skill registration
- MCP tools
- Executing ripple delete
- LLM-based filler classification
- Language-model rewriting
- Removal of arbitrary pauses or silence

## Safety and capability boundaries

- A cut must stay inside the requested target and occurrence.
- A cut must not overlap adjacent protected speech.
- Ambiguous mappings, overlapping speech, or missing VAD evidence fail closed.
- Low-confidence candidates are never `AUTO_APPLY`.
- The resolver must not call an editor adapter.
- Floating-point rounding must not move a boundary outside the safe rational range.

## Dependencies

- Depends on “Add revision-bound speech and VAD analysis.”
- Uses the closed-loop speech editing policy defined for v0.0.3.

## Acceptance criteria

- [ ] Candidate generation is deterministic for identical input.
- [ ] Candidate IDs remain stable for one preview input and revision.
- [ ] Decisions use `AUTO_APPLY`, `SUGGESTED`, and `SKIPPED`.
- [ ] Every decision includes structured evidence and reason codes.
- [ ] Safe ranges are aligned to sequence frame duration.
- [ ] Ranges never cross target or occurrence boundaries.
- [ ] Adjacent words and protected segments are preserved.
- [ ] Missing VAD, ambiguous mapping, overlapping speech, and low confidence fail closed.
- [ ] Valid candidates produce editor-independent `ripple-delete` operations.
- [ ] No editor adapter dependency is introduced into the runtime component.

## Validation

Add deterministic fixtures for obvious fillers, low confidence, adjacent words, clip boundaries, trimmed occurrences, overlapping speech, protected laughter and breath, short pauses, multiple fillers, and non-integral frame boundaries.

Add invariants proving that resolved ranges are non-overlapping, frame-aligned, bounded, and deterministic.

### PR DESCRIPTION
**Pull request by morshoto**
## Summary

Implemented the editor-independent runtime components for issue #42:

- Added `FillerDetector` with analyzer metadata, configurable vocabulary, confidence evidence, deterministic preview-scoped IDs, and source-to-occurrence mapping.
- Added `SafeCutResolver` with VAD/silence/protected-segment checks, fail-closed decisions, exact rational frame alignment, pause preservation, and pure `ripple-delete` operations.
- Extended fixture speech analysis to preserve range-bound VAD and protected evidence.
- Added deterministic unit/property-style coverage and documented the contract.

Closes: #42

## Validation

```bash
pnpm install --frozen-lockfile
pnpm run build
pnpm run test
pnpm run check:boundaries
```

All passed: 396 tests, build, and boundary checks. Native/Xcode validation is not applicable; no native files changed.

## Architecture and compatibility

- [x] Runtime remains independent of MCP and editor-specific implementations.
- [x] Public package interfaces are documented.
- [x] Existing adapters and test fixtures remain compatible.

## Safety

- [x] No credentials, private media, raw crash dumps, or user-specific local paths were added.
- [x] Generated files and build artifacts are excluded.
- [x] Documentation reflects the changed runtime contract.

## Summary by CodeRabbit

- **New Features**
  - Added deterministic detection of filler words in speech, with confidence scoring and clear evidence.
  - Added safe-cut resolution that preserves natural pauses, respects speech boundaries, and avoids protected content.
  - Added frame-aligned ripple-delete decisions with actionable statuses for accepted, suggested, or rejected cuts.
  - Added support for speech, silence, breath, laughter, and noise segment data.
  - Exposed filler detection and safe-cut capabilities through the runtime package.

- **Documentation**
  - Documented filler detection rules, safe-cut behavior, evidence, and decision statuses.

- **Tests**
  - Expanded automated coverage for speech analysis, detection, cut safety, and edge cases.

### CODE REVIEW

File: packages/runtime/src/speech/filler-detector.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    throw new Error(`INVALID_OPERATION: filler ${label} must be positive`);
+  }
+}
+
+function containsRange(container: TimeRange, candidate: TimeRange): boolean {
+  return candidate.start >= container.start && candidate.end <= container.end;
+}
+
+function normalizeWord(text: string): string {
+  return text.trim().toLocaleLowerCase().replace(/[.,!?;:]+$/u, "");
```
[the remark this answers is not in the document]
Comment (morshoto) [reply packages/runtime/src/speech/filler-detector.ts]: Fixed in 81f0d24. Replaced the trailing-punctuation regex with a bounded character scan. The repeated-punctuation regression passes; full local validation is green: build, 414 tests, and boundary checks.

File: packages/runtime/src/speech/safe-cut-resolver.ts [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+    const minimumSequenceStart = Math.max(
+      mapSourceToSequence(minimumSourceStart, request.occurrence),
+      request.targetRange.start,
+      request.occurrence.sequenceRange.start,
+    );
+    const maximumSequenceEnd = Math.min(
+      mapSourceToSequence(nextStart, request.occurrence),
+      request.targetRange.end,
+      request.occurrence.sequenceRange.end,
+    );
```
[the remark this answers is not in the document]
Comment (morshoto) [reply packages/runtime/src/speech/safe-cut-resolver.ts:197]: Fixed in a014ed8. Safe frame bounds now clamp against neighboring speech VAD boundaries as well as transcript words. Added regression coverage for untranscribed speech; full local validation is green: build, 414 tests, and boundary checks.

**Comment by morshoto** [comment]
Review status at f731ba3:

- Fixed in 4aaa187: candidate IDs no longer depend on analysis-window word index.
- Fixed in fbb7530: malformed silence and protected evidence have dedicated reason codes.
- Fixed in 2b11983: removed the unreachable redundant word-overlap condition.
- Fixed in a014ed8: frame bounds clamp against neighboring speech VAD boundaries.
- Fixed in 81f0d24: replaced the CodeQL-flagged polynomial regex with a bounded scan.
- Fixed in f731ba3: updated the filler detection and safe-cut contract documentation.
- Declined [info]: CodeRabbit docstring coverage suggestion; this repository has no docstring gate, lint/static checks pass, and no behavior gap was identified.
